opencog/pln/rules/intensional/intensional-difference-members-direct-introduction.scm

```diff
@@ -0,0 +1,113 @@
+;; Rule for introducing
+;;
+;; MemberLink
+;;   X
+;;   IntensionalDifferenceLink
+;;     A
+;;     B
+;;
+;; based on the direct evidence of patterns of A and B, and
+;; the intensional difference between them.
+;;
+;; IntensionalDifferenceLink
+;;   A
+;;   B
+;; AttractionLink
+;;   A
+;;   X
+;; AttractionLink
+;;   B
+;;   X
+;; |-
+;; MemberLink <TV>
+;;   X
+;;   IntensionalDifferenceLink
+;;     A
+;;     B
+
+;; Rule
+(define intensional-difference-members-direct-introduction-rule
+  (let* ((A (Variable "$A"))
+         (B (Variable "$B"))
+         (X (Variable "$X"))
+         (CT (Type "ConceptNode")))
+    (Bind
+      (VariableSet
+        (TypedVariable A CT)
+        (TypedVariable B CT)
+        (TypedVariable X CT))
+      (And
+        (Present
+          (IntensionalDifference A B)
+          (Attraction A X)
+          (Attraction B X)))
+      (ExecutionOutput
+        (GroundedSchema "scm: intensional-difference-members-direct-introduction")
+        (List
+          ;; Conclusion
+          (Member X (IntensionalDifference A B))
+          ;; Premise
+          (IntensionalDifference A B))))))
+
+;; Formula
+(define (intensional-difference-members-direct-introduction conclusion premise)
+  ;; Given two concepts A and B, return the attraction link
+  ;;
+  ;; Attraction <TV>
+  ;;   A
+  ;;   B
+  (define (get-attraction A B)
+    (cog-link 'AttractionLink A B))
+
+  ;; Given an attraction link, return the strength which is
+  ;; the product of the mean and the confidence of the TV on
+  ;; on the attraction link
+  ;;
+  ;; Attraction <TV>
+  ;;   A
+  ;;   B
+  (define (get-pattern-strength at)
+      (if (null? at) 0 (* (cog-mean at) (cog-confidence at))))
+
+  ;; Given the attraction links of A and B, calculate the
+  ;; strength of the TV expressed as
+  ;;
+  ;; min(attraction(A,X), 1 - attraction(B,X))
+  ;;
+  ;; where the attraction(A,X) is the strength of the TV of
+  ;;
+  ;; Attraction <TV>
+  ;;   A
+  ;;   X
+  (define (get-strength A-at B-at)
+    (min (cog-mean A-at) (- 1 (get-pattern-strength B-at))))
```

Comment:
This is OK, but I think you'll get vastly superior performance and more readable, easier-to-understand code if you use PredicateFormulas. See `/examples/atomspace/formulas.scm` for working examples.

Reply:
That's interesting @linas, I however don't understand how the TV of the conclusion is being set then, without invoking a grounded schema.

Reply:
Did you try running the example code?

Reply:
Yes, was able to set the TV on the `EvaluationLink` holding the `PredicateFormulaLink`, but what other atom types?

Reply:
> what other atom types?

I don't understand the question. Do you want to use GPN's (or DefinedPredNodes) with something other than evaluation links?

Reply:
For instance, given

```scheme
Inheritance A B
Inheritance B C
```

we want to calculate the TV of

```scheme
Inheritance A C
```

and set it on the inferred inheritance.

Reply:
Oh, gack. I just looked more carefully; I misunderstood the original use-case.   You are abusing an "undocumented behavior" of the GSN to perform a side-effect, with the side-effect being to alter the TV of some other completely unrelated atom, having nothing at all to do with the GSN.  Yuck.  I guess you got lucky that this code works. Its certainly not the intended use case, and is fairly abusive of the system. 

I don't particularly see any clean way of doing this.  This is yet another fundamental architectural issue that we've ignored, and just hacked-around instead, and its clearly biting in the heinie now.  Ugh. They are starting to pile up, each one unfixed...